drivers/photo-electronic/adpd410x/adpd410x.c

```diff
@@ -0,0 +1,660 @@
+
+	/* Set ADC cycle and repeat */
+	if(init->adc_cycles == 0)
+		init->adc_cycles = 1;
+	if(init->repeats_no == 0)
+		init->repeats_no = 1;
+	data = init->repeats_no | (init->adc_cycles << BITP_COUNTS_A_NUM_INT_A);
+	return adpd410x_reg_write(dev, (ADPD410X_REG_COUNTS_A + timeslot_no * 0x20),
+				  data);
+}
+
+/**
+ * @brief Get number of bytes in the device FIFO.
+ * @param dev - Device handler.
+ * @param bytes - Pointer to the byte count container.
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+int32_t adpd410x_get_fifo_bytecount(struct adpd410x_dev *dev, uint16_t *bytes)
+{
+	int32_t ret;
+
+	ret = adpd410x_reg_read(dev, ADPD410X_REG_FIFO_STATUS, bytes);
+	if(ret != SUCCESS)
+		return FAILURE;
+
+	*bytes &= BITM_INT_STATUS_FIFO_FIFO_BYTE_COUNT;
+
+	return SUCCESS;
+}
+
+/**
+ * @brief Get a data packet containing data from all active time slots and
+ *        channels. adpd410x_get_data() helper function.
+ * @param dev - Device handler.
+ * @param data - Pointer to the data container.
+ * @param no_slots - Number of active time slots.
+ * @param dual_chan - Mask showing which of the active time slots are dual
+ *                    channeled and which are single channeled.
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+static int32_t adpd410x_get_data_packet(struct adpd410x_dev *dev,
+					uint32_t *data, uint8_t no_slots, uint16_t dual_chan)
+{
+	int32_t ret;
+	uint16_t data_byte_count, temp_data, expect_byte_count = 0;
+	uint16_t *data_buff, sample_index = 0;
+	int8_t i, got_one = 0;
+	uint8_t *slot_bytecount_tab, *data_byte_buff;
+
+	slot_bytecount_tab = (uint8_t *)calloc(no_slots,
+					       sizeof (*slot_bytecount_tab));
+	if (!slot_bytecount_tab)
+		return FAILURE;
+	for(i = 0; i < no_slots; i++) {
+		ret = adpd410x_reg_read(dev, (ADPD410X_REG_DATA1_A + i * 0x20),
+					&temp_data);
+		if(ret != SUCCESS)
+			goto error_slot;
+		slot_bytecount_tab[i] = (temp_data &
+					 BITM_DATA1_A_SIGNAL_SIZE_A);
+		expect_byte_count += slot_bytecount_tab[i];
+		if((dual_chan & (1 << i)) != 0)
+			expect_byte_count += slot_bytecount_tab[i];
+	}
+	if((expect_byte_count % 2) != 0)
+		expect_byte_count++;
+	data_buff = (uint16_t *)calloc((expect_byte_count / 2),
+				       sizeof (*data_buff));
+	if (!data_buff)
+		goto error_slot;
+	data_byte_buff = (uint8_t *)calloc(expect_byte_count,
+					   sizeof (*data_byte_buff));
+	if (!data_byte_buff)
+		goto error_data_buff;
+
+	for(i = 0; i < (expect_byte_count / 2); i++) {
+		ret = adpd410x_reg_read(dev, ADPD410X_REG_FIFO_DATA,
+					(data_buff + i));
+		if(ret != SUCCESS)
+			goto error_data_byte;
+	}
+	for(i = 0; i < (expect_byte_count / 2); i++) {
+		data_byte_buff[(i * 2)] = (data_buff[i] & 0xff00) >> 8;
+		data_byte_buff[(i * 2 + 1)] = data_buff[i] & 0xff;
+	}
+
+	i = 0;
+	data_byte_count = 0;
+	do {
+		switch(slot_bytecount_tab[i]) {
+		case 0:
+			continue;
+		case 1:
+			data[sample_index] = data_byte_buff[data_byte_count];
+			data_byte_count++;
```

Comment:
```suggestion
			data[sample_index] = data_byte_buff[j++];
```
I think this is better than my other suggestion.
I think you can rename data_byte_count to j because it is just an index and increment it inside []